src/boson/bson_streambuf.cpp

```diff
@@ -67,7 +67,7 @@ int bson_output_streambuf::insert(int ch) {
     }
 
     // This creates the document from the given bytes, and calls the user-provided callback.
-    if (_bytes_read == _len) {
+    if (_bytes_read == _len && _len > 4) {
```

Comment:
How about earlier(e..g line 54)  just returning early if _bytes_read < 4, so we're never dealing with an invalid _len in the later code?